src/net/socket_ancillary.rs

```diff
@@ -0,0 +1,777 @@
+        unsafe { Uid::from_raw(self.0.uid) }
+    }
+
+    /// Set the GID.
+    pub fn set_gid(&mut self, gid: Gid) {
+        self.0.gid = gid.as_raw();
+    }
+
+    /// Get the current GID.
+    pub fn get_gid(&self) -> Gid {
+        unsafe { Gid::from_raw(self.0.gid) }
+    }
+}
+
+// TODO: Provide way of sizing the buffer for SocketAncillary upfront, like in
+// https://docs.rs/nix/latest/nix/macro.cmsg_space.html
+
+// TODO: Find a way to use MaybeUninit as backing data.
+
+// TODO: Should there exist a convenience wrapper that owns the buffer and potentially
+// auto resizes?
+
+// TODO: port tests from https://github.com/nix-rust/nix/blob/master/test/sys/test_socket.rs
+
+/// TODO: document
+#[derive(Debug)]
+pub struct SocketAncillary<'a, T: TryFrom<&'a c::cmsghdr>> {
+    pub(crate) buffer: &'a mut [u8],
+    pub(crate) length: usize,
+    pub(crate) truncated: bool,
+    _t: PhantomData<T>,
+}
+
+impl<'a, T: TryFrom<&'a c::cmsghdr>> SocketAncillary<'a, T> {
+    /// Create an ancillary data with the given buffer.
+    pub fn new(buffer: &'a mut [u8]) -> Self {
+        SocketAncillary {
+            buffer,
+            length: 0,
+            truncated: false,
+            _t: Default::default(),
+        }
+    }
+
+    /// Returns the capacity of the buffer.
+    pub fn capacity(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Returns `true` if the ancillary data is empty.
+    pub fn is_empty(&self) -> bool {
+        self.length == 0
+    }
+
+    /// Returns the number of used bytes.
+    pub fn len(&self) -> usize {
+        self.length
+    }
+
+    /// Is `true` if during a recv operation the ancillary was truncated.
+    pub fn truncated(&self) -> bool {
+        self.truncated
+    }
+
+    /// Clears the ancillary data, removing all values.
+    pub fn clear(&mut self) {
+        self.length = 0;
+        self.truncated = false;
+    }
+
+    /// Returns the iterator of the control messages.
+    pub fn messages<'b: 'a>(&'b self) -> Messages<'a, T> {
+        Messages {
+            buffer: &self.buffer[..self.length],
+            current: None,
+            _t: Default::default(),
+        }
+    }
+}
+
+/// TODO: document
+pub type SendSocketAncillaryUnix<'a> = SocketAncillary<'a, SendAncillaryDataUnix<'a>>;
+
+impl<'a> SocketAncillary<'a, SendAncillaryDataUnix<'a>> {
+    /// Add file descriptors to the ancillary data.
+    ///
+    /// The function returns `true` if there was enough space in the buffer.
+    /// If there was not enough space then no file descriptors was appended.
+    /// Technically, that means this operation adds a control message with the level `SOL_SOCKET`
+    /// and type `SCM_RIGHTS`.
+    pub fn add_fds<Fd: AsFd>(&mut self, fds: &[Fd]) -> bool {
+        self.truncated = false;
+        add_to_ancillary_data(
+            &mut self.buffer,
+            &mut self.length,
+            fds,
+            c::SOL_SOCKET as _,
+            c::SCM_RIGHTS as _,
+        )
```

Comment:
`add_to_ancillary_data` does a raw `copy_nonoverlapping` from the source (`fds` here) into the message buffer. That might work for types like `File` or `TcpStream` which only contain a file descriptor and nothing else, but here, all we know is that `Fd: AsFd`, so `Fd` could contain other state. It looks like something here needs to iterate over the `fds` and call `as_fd().as_raw_fd()` on them individually to ensure that we're only passing file-descriptor values and not other arbitrary data here.

Reply:
fixed